### PR TITLE
Default tags

### PR DESCRIPTION
**Pull request by danielrbradley**
Fixes https://github.com/pulumi/pulumi-aws-native/issues/107

Implement the `defaultTags` provider configuration option. When setting tags on a resource, match the tags style as defined in the metadata.

### CODE REVIEW

File: provider/pkg/schema/gen.go [hunk 1]
```diff
@@ -851,6 +865,16 @@ func (ctx *context) gatherResourceType() error {
 		if err != nil {
 			return err
 		}
+		if hasTags && prop == tagsProp {
```
Comment (thomas11) [inline provider/pkg/schema/gen.go:868]: This added block seems to be identical to the one above? Can we extract it?

File: provider/pkg/schema/gen_tags.go [hunk 2]
```diff
@@ -29,24 +31,62 @@ const (
...
-	}
-	asInterface, ok := tagging.(map[string]interface{})
-	if !ok {
-		return "", false
+func (ts TagsStyle) IsStringMap() bool {
+	return ts == TagsStyleStringMap
+}
+
+func (ts TagsStyle) IsKeyValueArray() bool {
+	return strings.HasPrefix(string(ts), "keyValueArray")
```
Comment (thomas11) [inline provider/pkg/schema/gen_tags.go:39]: Would be a tad safer to use a `const keyValueArrayPrefix` here and in the names

File: provider/pkg/schema/gen_tags.go [hunk 3]
```diff
@@ -29,24 +31,62 @@ const (
...
+	default: // Unknown
+		ctx.reports.UnexpectedTagsShapes[ctx.resourceToken] = spec
 	}
-	tagProperty, ok := asInterface["tagProperty"]
-	if !ok {
-		return "", false
+}
+
+func GetTagsProperty(originalSpec *jsschema.Schema) (string, bool) {
+	fromTaggingMetadata, ok := func() (string, bool) {
```
Comment (thomas11) [inline provider/pkg/schema/gen_tags.go:63]: What's the reason for using an inline function that's instantly run? Isn't that the same as inlining the function body?

File: provider/pkg/provider/provider.go [hunk 4]
```diff
@@ -459,6 +457,17 @@ func (p *cfnProvider) Configure(ctx context.Context, req *pulumirpc.ConfigureReq
 		}
 	}
 
+	if defaultTagsJson, ok := vars["aws-native:config:defaultTags"]; ok {
+		var defaultTags map[string]string
+		err := json.Unmarshal([]byte(defaultTagsJson), &defaultTags)
+		if err != nil {
+			return nil, fmt.Errorf("failed to unmarshal 'forbiddenAccountIds' config: %w", err)
```
Comment (thomas11) [inline provider/pkg/provider/provider.go:464]: wrong copied name

File: provider/pkg/provider/provider.go [hunk 5]
```diff
@@ -459,6 +457,17 @@ func (p *cfnProvider) Configure(ctx context.Context, req *pulumirpc.ConfigureReq
...
+		}
+		p.defaultTags = defaultTags
+	} else {
+		p.defaultTags = make(map[string]string)
```
Comment (thomas11) [inline provider/pkg/provider/provider.go:468]: nil maps are usable for lookups, do we need an actual map?

File: provider/pkg/provider/provider_defaultTags_test.go [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+import (
+	"testing"
+
+	"github.com/pulumi/pulumi-aws-native/provider/pkg/schema"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestDefaultTags(t *testing.T) {
```
Comment (thomas11) [inline provider/pkg/provider/provider_defaultTags_test.go:11]: There are some more cases that would be good to cover IMO, like not overwriting existing tags with default tags.

File: provider/pkg/provider/provider.go (hunk 5 above)
Comment (danielrbradley) [reply provider/pkg/provider/provider.go:468]: Sure, I'll change to a nil assignment. It's kinda an edge case, but when debugging the provider, if we don't allocate a nil here it causes a weird bug when we remove the default tags from the config as the old value will then remain set if we don't explicitly unset it.

File: provider/pkg/schema/gen.go (hunk 1 above)
Comment (danielrbradley) [reply provider/pkg/schema/gen.go:861]: Done 👍

**Review by thomas11 — APPROVED** [review]

**Comment by fitz-vivodyne** [comment]
Is this supported now?
The [documentation](https://www.pulumi.com/registry/packages/aws-native/installation-configuration/#configuration-options) still states that it isn't

**Comment by danielrbradley** [comment]
Yes, this works in v0.100.0

**Pull request merged**